receiver/awscontainerinsightreceiver/internal/efa/efaSysfs.go

```diff
@@ -0,0 +1,402 @@
+)
+
+const (
+	defaultCollectionInterval = 20 * time.Second
+)
+
+const (
+	efaPath            = "/sys/class/infiniband"
+	efaK8sResourceName = "vpc.amazonaws.com/efa"
+
+	// hardware counter names
+	counterRdmaReadBytes      = "rdma_read_bytes"
+	counterRdmaWriteBytes     = "rdma_write_bytes"
+	counterRdmaWriteRecvBytes = "rdma_write_recv_bytes"
+	counterRxBytes            = "rx_bytes"
+	counterRxDrops            = "rx_drops"
+	counterTxBytes            = "tx_bytes"
+)
+
+var counterNames = map[string]any{
+	counterRdmaReadBytes:      nil,
+	counterRdmaWriteBytes:     nil,
+	counterRdmaWriteRecvBytes: nil,
+	counterRxBytes:            nil,
+	counterRxDrops:            nil,
+	counterTxBytes:            nil,
+}
+
+type Scraper struct {
+	collectionInterval time.Duration
+	cancel             context.CancelFunc
+
+	sysFsReader       sysFsReader
+	deltaCalculator   metrics.MetricCalculator
+	decorator         stores.Decorator
+	podResourcesStore podResourcesStore
+	store             *efaStore
+	logger            *zap.Logger
+}
+
+type sysFsReader interface {
+	ListDevices() ([]efaDeviceName, error)
+	ListPorts(deviceName efaDeviceName) ([]string, error)
+	ReadCounter(deviceName efaDeviceName, port string, counter string) (uint64, error)
+}
+
+type podResourcesStore interface {
+	AddResourceName(resourceName string)
+	GetContainerInfo(deviceID string, resourceName string) *stores.ContainerInfo
+}
+
+type efaStore struct {
+	timestamp time.Time
+	devices   *efaDevices
+}
+
+// efaDevices is a collection of every Amazon Elastic Fabric Adapter (EFA) device in
+// /sys/class/infiniband.
+type efaDevices map[efaDeviceName]*efaCounters
+
+type efaDeviceName string
+
+// efaCounters contains counter values from files in
+// /sys/class/infiniband/<Name>/ports/<Port>/hw_counters
+// for a single port of one Amazon Elastic Fabric Adapter device.
+type efaCounters struct {
+	rdmaReadBytes      uint64 // hw_counters/rdma_read_bytes
+	rdmaWriteBytes     uint64 // hw_counters/rdma_write_bytes
+	rdmaWriteRecvBytes uint64 // hw_counters/rdma_write_recv_bytes
+	rxBytes            uint64 // hw_counters/rx_bytes
+	rxDrops            uint64 // hw_counters/rx_drops
+	txBytes            uint64 // hw_counters/tx_bytes
+}
+
+func NewEfaSyfsScraper(logger *zap.Logger, decorator stores.Decorator, podResourcesStore podResourcesStore) *Scraper {
+	ctx, cancel := context.WithCancel(context.Background())
+	podResourcesStore.AddResourceName(efaK8sResourceName)
+	e := &Scraper{
+		collectionInterval: defaultCollectionInterval,
+		cancel:             cancel,
+		sysFsReader:        defaultSysFsReader,
+		deltaCalculator:    metrics.NewMetricCalculator(calculateDelta),
+		decorator:          decorator,
+		podResourcesStore:  podResourcesStore,
+		store:              new(efaStore),
+		logger:             logger,
+	}
+
+	go e.startScrape(ctx)
+
+	return e
+}
+
+func calculateDelta(prev *metrics.MetricValue, val any, _ time.Time) (any, bool) {
+	var deltaValue uint64
+	if prev != nil {
```

Comment:
nit: probably return the default(0) with false when it's null upfront. otherwise:
`return val.(uint64) - prev.RawValue.(uint64), true`

Reply:
Thanks, had copied this from somewhere else without thinking about it.  Your suggestion is much nicer.